### ISSUE TITLE
[Refactor] Select ML Circuit Breaking Functionality to core library

### ISSUE THREAD

**Issue by nknize** [opensearch-project/ml-commons#1171]
**Is your feature request related to a problem?**
ML-Commons provides some useful top level circuit breakers that could add value to the core library (e.g., `DiskCircuitBreaker`). However, it also duplicates core circuit breaking functionality (e.g., `MemoryCircuitBreaker`) which causes confusion for those taking a dependency on both the `:libs:opensearch-core` and `ml-commons` libraries. In addition, creating a [circuit breaker service outside the core](https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/main/java/org/opensearch/ml/breaker/MLCircuitBreakerService.java) (without taking a dependency on the core service mechanics) bypasses a lot of additional memory tracking functionality provided by the core (e.g., [Stats tracking](https://github.com/opensearch-project/OpenSearch/blob/0b1f875d34c7d723afd41396234f8f2003c6be2c/server/src/main/java/org/opensearch/indices/breaker/CircuitBreakerStats.java#L50), and [Durability](https://github.com/opensearch-project/OpenSearch/blob/0b1f875d34c7d723afd41396234f8f2003c6be2c/server/src/main/java/org/opensearch/common/breaker/CircuitBreaker.java#L104)) which is useful in Node resource tracking and communicating memory usaage up to sys admins and support engineers.  

**What solution would you like?**
Converge the circuit breaking implementation with the core. We should refactor the `DiskCircuitBreaker` to the core so we can use it in other places (e.g., preemptive Disk Watermark checks) and enable the ml-commons breakers to leverage the memory and resource usage stats already tracked by the core.

**What alternatives have you considered?**
Leave the fragmentation as is...

**Comment by ylwu-amzn** [comment]
Thanks @nknize. I think this type of fragmentation exists in multiple plugins. +1 for building more types of circuit breaker in core.

**Comment by saratvemulapalli** [comment]
+1 @nknize. I've been looking at ML-Commons lately there are few pieces which are built as custom for the plugin which should really be part of core as a library.
- Circuit Breakers
- Task Manager 
- Distributing resources in a cluster
- Request router

**Comment by nknize** [comment]
> * Circuit Breakers
> * Task Manager

Awesome. I have [an open PR](https://github.com/opensearch-project/OpenSearch/pull/9006) for refactoring the Circuit Breaker foundation to the core library. And two followup PRs coming for refactoring the `Settings` and `TaskManager` foundation to the core library. 
Maybe we open a core `Meta` issue for refactoring these four mechanisms from ml-commons to a new core `:libs:opensearch-ml` library?

**Comment by ylwu-amzn** [comment]
Thanks @nknize .

I had a offline conversation with @saratvemulapalli , the ML task manager has something in common and something special for ML. For example, the task type is highly coupled with ML. But the design also has some common part. I think it can benefit some other application if copy that common part over to core. The same for distributing resources and request router (we call it task dispatcher in ml-commons). Should be good to copy the common part to core.

**Comment by ylwu-amzn** [comment]
One suggestion: to make the discussion easier, let's keep this issue for circuit breaking functionality only. Create separate issues for other topics.

**Comment by saratvemulapalli** [comment]
> One suggestion: to make the discussion easier, let's keep this issue for circuit breaking functionality only. Create separate issues for other topics.

Sure, I'll probably open another issue for all of them together.

**Comment by nknize** [comment]
> For example, the task type is highly coupled with ML.

So this is part of the motivation for #1172. If we work toward providing core ML functionality (e.g., kmeans trainer) as a module in the core min distribution then we can move the ML specific TaskType logic to a new `:libs:opensearch-ml` library and leverage it in a new `:module:core-ml` that also provides the `MLEngine` for downstream repos (e.g., this one) to extend w/ more specialized fine tuned training logic (including GPU accelleration). This would enable other users to write their own trainers as plugins (or extensions) while providing an OTB ML capability as part of the min distribution.

**Comment by ylwu-amzn** [comment]
> So this is part of the motivation for https://github.com/opensearch-project/ml-commons/issues/1172. If we work toward providing core ML functionality (e.g., kmeans trainer) as a module in the core min distribution then we can move the ML specific TaskType logic to a new :libs:opensearch-ml library and leverage it in a new :module:core-ml that also provides the MLEngine for downstream repos (e.g., this one) to extend w/ more specialized fine tuned training logic (including GPU accelleration). This would enable other users to write their own trainers as plugins (or extensions) while providing an OTB ML capability as part of the min distribution.

This is why I said what your proposed is not just port Kmeans to core, it's going to port the whole ML framework. This will change our roadmap totally.

**Comment by nknize** [comment]
> This will change our roadmap totally.

hrm... What specifically will change? Maybe there's a project board or github issue you can post here that outlines the roadmap that will help single out the specific features that would be blocked by this convergence?

**Comment by ylwu-amzn** [comment]
Like I said moving the Kmeans to core will also move the whole ML framework. I won't say this will block our future feature development. But moving these things around will add some effort and slow down the pace. I guess the main point moving to core repo is to make things extensible. Seems another option if we can achieve the same goal by fine tuning current ml-commons.

**Comment by ylwu-amzn** [comment]
Remove `untriaged` as we all agree to build more general circuit breaker in core.
